microovn/ovn/certificates.go

```diff
@@ -0,0 +1,251 @@
+package ovn
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"database/sql"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"math/big"
+	"os"
+	"time"
+
+	"github.com/canonical/microcluster/state"
+	"github.com/canonical/microovn/microovn/database"
+	"github.com/canonical/microovn/microovn/ovn/paths"
+)
+
+const CACertName = "cacert"
+const CACertValidity = 10 * 365 * 24 * time.Hour     // 10 years
+const ServiceCertValidity = 2 * 365 * 24 * time.Hour // 2 years
+
+var MaxSerialNumber = new(big.Int).Lsh(big.NewInt(1), 128)
+
+type CertificateType int
+
+const (
+	CertificateTypeCA CertificateType = iota
+	CertificateTypeServer
+	CertificateTypeClient
+)
+
+// issueCertificate function generates new "server", "client" or "CA" certificate based on the value passed
+// to the "certType" argument. Argument "cn" is passed unchanged to the certificate's CN and "serviceName" is used in
+// certificate's OU.  Certificates are valid, based on their type, for a period specified in CACertValidity
+// and ServiceCertValidity.
+//
+// When generating certificate that is signed by a CA, "parent" argument must point to a parsed CA certificate and
+// "signer" argument must point to CA's private key. On the other hand if you want to generate self-signed certificate,
+// both "parent" and "signer" arguments must be empty (nil).
+//
+// This function returns PEM encoded certificate, private key and error (if any occurred).
+func issueCertificate(cn string, serviceName string, certType CertificateType, parent *x509.Certificate, signer *ecdsa.PrivateKey) ([]byte, []byte, error) {
+	var (
+		isCa     bool
+		keyUsage x509.KeyUsage
+		signKey  *ecdsa.PrivateKey
+		validTo  time.Time
+	)
+	// Generate certificate's private key
+	keyPair, err := ecdsa.GenerateKey(elliptic.P384(), rand.Reader)
```

Comment:
What is the reasoning behind this choice of key material?

Reply:
I took again inspiration from [lxd](https://github.com/lxc/lxd/blob/123231d27121223f20f11a65499ce8a5e35d704b/shared/cert.go#L311), and I think the choice is fine. I believe that there's little difference to us whether we go with curves or RSA, aside from key sizes, as both approaches are considered secure. I don't think we need to be too concerned about whether NSA has backdoors in ECDSA curves.
Now that I think about it more, maybe we can take it down a notch and go with `p-256` instead of `p-384` for better performance and still acceptable level of security, what do you think? I honestly didn't dwell to much on this choice as the aim is not to have perfectly secure system here.

Reply:
No need to change it, I just wanted to know how we came to this conclusion. Adding a comment about why we selected this specific key material is in order, either a reference to LXD or a reference to why they ended up with that selection.